internal/dispatch/dispatch.go

```diff
@@ -65,9 +65,21 @@ func CheckDepth(ctx context.Context, req HasMetadata) error {
 	return nil
 }
 
-// CheckRequestToKey converts a check request into a cache key
+// CheckRequestToKey converts a check request into a cache key based on the relation
 func CheckRequestToKey(req *v1.DispatchCheckRequest) string {
-	return fmt.Sprintf("check//%s@%s@%s", tuple.StringONR(req.ObjectAndRelation), tuple.StringONR(req.Subject), req.Metadata.AtRevision)
+	return fmt.Sprintf("check//relation/%s@%s@%s", tuple.StringONR(req.ObjectAndRelation), tuple.StringONR(req.Subject), req.Metadata.AtRevision)
+}
+
+// CheckRequestToKeyWithPossibleCanonical converts a check request into a cache key based possibly
+// on the canonical key, or the relation name if the canonical key is empty
+func CheckRequestToKeyWithPossibleCanonical(req *v1.DispatchCheckRequest, possibleCanonicalKey string) string {
```

Comment:
They can't have the same key as there is a non-zero possibility someone creates a relation with a name that matches the computed hash, which would result in the cache keys overlapping incorrectly. I explicitly have different prefixes to prevent that from occurring

Reply:
We always compute the cache keys ourselves ahead of time, if that's a concern why not just detect it and either reject the schema or change the canonical key?

Reply:
Two reasons:
(1) Its not backwards compatible and limiting the names of relations to prevent overlap is a user facing error for an internal implementation
(2) Why do so when we can simply just construct a different key to *guarantee* no overlap? The cache itself is still shared

Reply:
Changing the canonical key if there's an existing relation with that hash (which I would be surprised if anyone, ever, hits) would also guarantee no overlap

we can pick a deterministic way to avoid conflicts: if hash(rudd) is the same as some user-provided name, calculate hash(hash(rudd)), etc

Reply:
Yes, but if we *ever* mess that up, then SpiceDB will return incorrect results. This *guarantees* there is no overlap, with essentially zero cost beyond the conditional to see if a canonical key exists at all

Reply:
just hash the relation names if we don't have a canonical key for it?

Reply:
I mean, I can trust that BDD won't collide its own hashed, but beyond that, I wouldn't want to risk it

Reply:
We're hashing the bdd output ourselves, and it's fnv which is much more likely to collide than a crypto hash fn

Reply:
Yeah, but it is isolated to the BDD input, vs the relation and BDD. I very much do not want to mix them. Why the push to have a single cache key?

Reply:
The "upside" of the design is to remove a conditional (?) and yet the downsides are:

1) If we ever mess it up, the cache will return *incorrect results*
2) If there is a collision, we either have to issue an error to the user (bad) or pick a new revision (which could completely cascade forward)

I really don't see any upside of real value and major downsides in additional complexity at annotation time OR user error... plus, we'll still need the conditional *anyway* because we won't necessarily have a cache key for older schemas anyway